src/lib/Imp.hs

```diff
@@ -51,10 +51,10 @@ import qualified Algebra as A
 -- TODO: use `Env ()` instead of `Scope`. The problem is that we're mixing up
 -- Imp and Core types in there.
 type EmbedEnv = ((Env Dest, [IVar]), (Scope, ImpProg))
-type ImpM = Cat EmbedEnv
+type ImpM = ReaderT Bindings (Cat EmbedEnv)
 
-toImpFunction :: ([ScalarTableVar], Block) -> (ImpFunction, Atom)
-toImpFunction (vsIn, block) = runImpM vsIn $ do
+toImpFunction :: Bindings -> ([ScalarTableVar], Block) -> (ImpFunction, Atom)
```

Comment:
What are the bindings here? Do they describe the data declarations? How about using a type alias to make it clear what they represent?

Reply:
Yes, they describe the data declarations. I don't think they're used any more, now that the data declarations are cached inline.